File: jscomp/stdlib/weak.ml
(***********************************************************************)
(*                                                                     *)
(*                                OCaml                                *)
(*                                                                     *)
(*            Damien Doligez, projet Para, INRIA Rocquencourt          *)
(*                                                                     *)
(*  Copyright 1997 Institut National de Recherche en Informatique et   *)
(*  en Automatique.  All rights reserved.  This file is distributed    *)
(*  under the terms of the GNU Library General Public License, with    *)
(*  the special exception on linking described in file ../LICENSE.     *)
(*                                                                     *)
(***********************************************************************)

(** Weak array operations *)

type 'a t;;

external create: int -> 'a t = "caml_weak_create";;

let length x = Obj.size(Obj.repr x) - 1;;

external set : 'a t -> int -> 'a option -> unit = "caml_weak_set";;
external get: 'a t -> int -> 'a option = "caml_weak_get";;
external get_copy: 'a t -> int -> 'a option = "caml_weak_get_copy";;
external check: 'a t -> int -> bool = "caml_weak_check";;
external blit: 'a t -> int -> 'a t -> int -> int -> unit = "caml_weak_blit";;
(* blit: src srcoff dst dstoff len *)

let fill ar ofs len x =
  if ofs < 0 || len < 0 || ofs + len > length ar
  then raise (Invalid_argument "Weak.fill")
  else begin
    for i = ofs to (ofs + len - 1) do
      set ar i x
    done
  end
;;

(** Weak hash tables *)

module type S = sig
  type data
  type t
  val create : int -> t
  val clear : t -> unit
  val merge : t -> data -> data
  val add : t -> data -> unit
  val remove : t -> data -> unit
  val find : t -> data -> data
  val find_all : t -> data -> data list
  val mem : t -> data -> bool
  val iter : (data -> unit) -> t -> unit
  val fold : (data -> 'a -> 'a) -> t -> 'a -> 'a
  val count : t -> int
  val stats : t -> int * int * int * int * int * int
end;;

module Make (H : Hashtbl.HashedType) : (S with type data = H.t) = struct

  type 'a weak_t = 'a t;;
  let weak_create = create;;
  let emptybucket = weak_create 0;;

  type data = H.t;;

  type t = {
    mutable table : data weak_t array;
    mutable hashes : int array array;
    mutable limit : int;               (* bucket size limit *)
    mutable oversize : int;            (* number of oversize buckets *)
    mutable rover : int;               (* for internal bookkeeping *)
  };;

  let get_index t h = (h land max_int) mod (Array.length t.table);;

  let limit = 7;;
  let over_limit = 2;;

  let create sz =
    let sz = if sz < 7 then 7 else sz in
    let sz = if sz > Sys.max_array_length then Sys.max_array_length else sz in
    {
      table = Array.make sz emptybucket;
      hashes = Array.make sz [| |];
      limit = limit;
      oversize = 0;
      rover = 0;
    };;

  let clear t =
    for i = 0 to Array.length t.table - 1 do
      t.table.(i) <- emptybucket;
      t.hashes.(i) <- [| |];
    done;
    t.limit <- limit;
    t.oversize <- 0;
  ;;

  let fold f t init =
    let rec fold_bucket i b accu =
      if i >= length b then accu else
      match get b i with
      | Some v -> fold_bucket (i+1) b (f v accu)
      | None -> fold_bucket (i+1) b accu
    in
    Array.fold_right (fold_bucket 0) t.table init
  ;;

  let iter f t =
    let rec iter_bucket i b =
      if i >= length b then () else
      match get b i with
      | Some v -> f v; iter_bucket (i+1) b
      | None -> iter_bucket (i+1) b
    in
    Array.iter (iter_bucket 0) t.table
  ;;

  let iter_weak f t =
    let rec iter_bucket i j b =
      if i >= length b then () else
      match check b i with
      | true -> f b t.hashes.(j) i; iter_bucket (i+1) j b
      | false -> iter_bucket (i+1) j b
    in
    Array.iteri (iter_bucket 0) t.table
  ;;

  let rec count_bucket i b accu =
    if i >= length b then accu else
    count_bucket (i+1) b (accu + (if check b i then 1 else 0))
  ;;

  let count t =
    Array.fold_right (count_bucket 0) t.table 0
  ;;

  let next_sz n = min (3 * n / 2 + 3) Sys.max_array_length;;
  let prev_sz n = ((n - 3) * 2 + 2) / 3;;

  let test_shrink_bucket t =
    let bucket = t.table.(t.rover) in
    let hbucket = t.hashes.(t.rover) in
    let len = length bucket in
    let prev_len = prev_sz len in
    let live = count_bucket 0 bucket 0 in
    if live <= prev_len then begin
      let rec loop i j =
        if j >= prev_len then begin
          if check bucket i then loop (i + 1) j
          else if check bucket j then begin
            blit bucket j bucket i 1;
            hbucket.(i) <- hbucket.(j);
            loop (i + 1) (j - 1);
          end else loop i (j - 1);
        end;
      in
      loop 0 (length bucket - 1);
      if prev_len = 0 then begin
        t.table.(t.rover) <- emptybucket;
        t.hashes.(t.rover) <- [| |];
      end else begin
        Obj.truncate (Obj.repr bucket) (prev_len + 1);
        Obj.truncate (Obj.repr hbucket) prev_len;
      end;
      if len > t.limit && prev_len <= t.limit then t.oversize <- t.oversize - 1;
    end;
    t.rover <- (t.rover + 1) mod (Array.length t.table);
  ;;

  let rec resize t =
    let oldlen = Array.length t.table in
    let newlen = next_sz oldlen in
    if newlen > oldlen then begin
      let newt = create newlen in
      let add_weak ob oh oi =
        let setter nb ni _ = blit ob oi nb ni 1 in
        let h = oh.(oi) in
        add_aux newt setter None h (get_index newt h);
      in
      iter_weak add_weak t;
      t.table <- newt.table;
      t.hashes <- newt.hashes;
      t.limit <- newt.limit;
      t.oversize <- newt.oversize;
      t.rover <- t.rover mod Array.length newt.table;
    end else begin
      t.limit <- max_int;             (* maximum size already reached *)
      t.oversize <- 0;
    end

  and add_aux t setter d h index =
    let bucket = t.table.(index) in
    let hashes = t.hashes.(index) in
    let sz = length bucket in
    let rec loop i =
      if i >= sz then begin
        let newsz = min (3 * sz / 2 + 3) (Sys.max_array_length - 1) in
        if newsz <= sz then failwith "Weak.Make: hash bucket cannot grow more";
        let newbucket = weak_create newsz in
        let newhashes = Array.make newsz 0 in
        blit bucket 0 newbucket 0 sz;
        Array.blit hashes 0 newhashes 0 sz;
        setter newbucket sz d;
        newhashes.(sz) <- h;
        t.table.(index) <- newbucket;
        t.hashes.(index) <- newhashes;
        if sz <= t.limit && newsz > t.limit then begin
          t.oversize <- t.oversize + 1;
          for _i = 0 to over_limit do test_shrink_bucket t done;
        end;
        if t.oversize > Array.length t.table / over_limit then resize t;
      end else if check bucket i then begin
        loop (i + 1)
      end else begin
        setter bucket i d;
        hashes.(i) <- h;
      end;
    in
    loop 0;
  ;;

  let add t d =
    let h = H.hash d in
    add_aux t set (Some d) h (get_index t h);
  ;;

  let find_or t d ifnotfound =
    let h = H.hash d in
    let index = get_index t h in
    let bucket = t.table.(index) in
    let hashes = t.hashes.(index) in
    let sz = length bucket in
    let rec loop i =
      if i >= sz then ifnotfound h index
      else if h = hashes.(i) then begin
        match get_copy bucket i with
        | Some v when H.equal v d
           -> begin match get bucket i with
              | Some v -> v
              | None -> loop (i + 1)
              end
        | _ -> loop (i + 1)
      end else loop (i + 1)
    in
    loop 0
  ;;

  let merge t d =
    find_or t d (fun h index -> add_aux t set (Some d) h index; d)
  ;;

  let find t d = find_or t d (fun h index -> raise Not_found);;

  let find_shadow t d iffound ifnotfound =
    let h = H.hash d in
    let index = get_index t h in
    let bucket = t.table.(index) in
    let hashes = t.hashes.(index) in
    let sz = length bucket in
    let rec loop i =
      if i >= sz then ifnotfound
      else if h = hashes.(i) then begin
        match get_copy bucket i with
        | Some v when H.equal v d -> iffound bucket i
        | _ -> loop (i + 1)
      end else loop (i + 1)
    in
    loop 0
  ;;

  let remove t d = find_shadow t d (fun w i -> set w i None) ();;

  let mem t d = find_shadow t d (fun w i -> true) false;;

  let find_all t d =
    let h = H.hash d in
    let index = get_index t h in
    let bucket = t.table.(index) in
    let hashes = t.hashes.(index) in
    let sz = length bucket in
    let rec loop i accu =
      if i >= sz then accu
      else if h = hashes.(i) then begin
        match get_copy bucket i with
        | Some v when H.equal v d
           -> begin match get bucket i with
              | Some v -> loop (i + 1) (v :: accu)
              | None -> loop (i + 1) accu
              end
        | _ -> loop (i + 1) accu
      end else loop (i + 1) accu
    in
    loop 0 []
  ;;

  let stats t =
    let len = Array.length t.table in
    let lens = Array.map length t.table in
    Array.sort compare lens;
    let totlen = Array.fold_left ( + ) 0 lens in
    (len, count t, totlen, lens.(0), lens.(len/2), lens.(len-1))
  ;;

end;;

File: lib/es6/weak.js


import * as Sys from "./sys.js";
import * as $$Array from "./array.js";
import * as Curry from "./curry.js";
import * as Caml_obj from "./caml_obj.js";
import * as Caml_weak from "./caml_weak.js";
import * as Caml_array from "./caml_array.js";
import * as Caml_int32 from "./caml_int32.js";
import * as Pervasives from "./pervasives.js";
import * as Caml_option from "./caml_option.js";
import * as Caml_primitive from "./caml_primitive.js";

function fill(ar, ofs, len, x) {
  if (ofs < 0 || len < 0 || (ofs + len | 0) > ar.length) {
    throw {
          ExceptionID: "Invalid_argument",
          _1: "Weak.fill"
        };
  }
  for(var i = ofs ,i_finish = ofs + len | 0; i < i_finish; ++i){
    Caml_weak.caml_weak_set(ar, i, x);
  }
  
}

function Make(H) {
  var emptybucket = Caml_weak.caml_weak_create(0);
  var get_index = function (t, h) {
    return (h & Pervasives.max_int) % t.table.length;
  };
  var create = function (sz) {
    var sz$1 = sz < 7 ? 7 : sz;
    var sz$2 = sz$1 > Sys.max_array_length ? Sys.max_array_length : sz$1;
    return {
            table: Caml_array.caml_make_vect(sz$2, emptybucket),
            hashes: Caml_array.caml_make_vect(sz$2, []),
            limit: 7,
            oversize: 0,
            rover: 0
          };
  };
  var clear = function (t) {
    for(var i = 0 ,i_finish = t.table.length; i < i_finish; ++i){
      Caml_array.caml_array_set(t.table, i, emptybucket);
      Caml_array.caml_array_set(t.hashes, i, []);
    }
    t.limit = 7;
    t.oversize = 0;
    
  };
  var fold = function (f, t, init) {
    return $$Array.fold_right((function (param, param$1) {
                  var _i = 0;
                  var _accu = param$1;
                  while(true) {
                    var accu = _accu;
                    var i = _i;
                    if (i >= param.length) {
                      return accu;
                    }
                    var v = Caml_weak.caml_weak_get(param, i);
                    if (v !== undefined) {
                      _accu = Curry._2(f, Caml_option.valFromOption(v), accu);
                      _i = i + 1 | 0;
                      continue ;
                    }
                    _i = i + 1 | 0;
                    continue ;
                  };
                }), t.table, init);
  };
  var iter = function (f, t) {
    return $$Array.iter((function (param) {
                  var _i = 0;
                  while(true) {
                    var i = _i;
                    if (i >= param.length) {
                      return ;
                    }
                    var v = Caml_weak.caml_weak_get(param, i);
                    if (v !== undefined) {
                      Curry._1(f, Caml_option.valFromOption(v));
                      _i = i + 1 | 0;
                      continue ;
                    }
                    _i = i + 1 | 0;
                    continue ;
                  };
                }), t.table);
  };
  var iter_weak = function (f, t) {
    return $$Array.iteri((function (param, param$1) {
                  var _i = 0;
                  while(true) {
                    var i = _i;
                    if (i >= param$1.length) {
                      return ;
                    }
                    if (Caml_weak.caml_weak_check(param$1, i)) {
                      Curry._3(f, param$1, Caml_array.caml_array_get(t.hashes, param), i);
                      _i = i + 1 | 0;
                      continue ;
                    }
                    _i = i + 1 | 0;
                    continue ;
                  };
                }), t.table);
  };
  var count_bucket = function (_i, b, _accu) {
    while(true) {
      var accu = _accu;
      var i = _i;
      if (i >= b.length) {
        return accu;
      }
      _accu = accu + (
        Caml_weak.caml_weak_check(b, i) ? 1 : 0
      ) | 0;
      _i = i + 1 | 0;
      continue ;
    };
  };
  var count = function (t) {
    return $$Array.fold_right((function (param, param$1) {
                  return count_bucket(0, param, param$1);
                }), t.table, 0);
  };
  var next_sz = function (n) {
    return Caml_primitive.caml_int_min((Caml_int32.imul(3, n) / 2 | 0) + 3 | 0, Sys.max_array_length);
  };
  var prev_sz = function (n) {
    return (((n - 3 | 0) << 1) + 2 | 0) / 3 | 0;
  };
  var test_shrink_bucket = function (t) {
    var bucket = Caml_array.caml_array_get(t.table, t.rover);
    var hbucket = Caml_array.caml_array_get(t.hashes, t.rover);
    var len = bucket.length;
    var prev_len = prev_sz(len);
    var live = count_bucket(0, bucket, 0);
    if (live <= prev_len) {
      var loop = function (_i, _j) {
        while(true) {
          var j = _j;
          var i = _i;
          if (j < prev_len) {
            return ;
          }
          if (Caml_weak.caml_weak_check(bucket, i)) {
            _i = i + 1 | 0;
            continue ;
          }
          if (Caml_weak.caml_weak_check(bucket, j)) {
            Caml_weak.caml_weak_blit(bucket, j, bucket, i, 1);
            Caml_array.caml_array_set(hbucket, i, Caml_array.caml_array_get(hbucket, j));
            _j = j - 1 | 0;
            _i = i + 1 | 0;
            continue ;
          }
          _j = j - 1 | 0;
          continue ;
        };
      };
      loop(0, bucket.length - 1 | 0);
      if (prev_len === 0) {
        Caml_array.caml_array_set(t.table, t.rover, emptybucket);
        Caml_array.caml_array_set(t.hashes, t.rover, []);
      } else {
        Caml_obj.caml_obj_truncate(bucket, prev_len + 0 | 0);
        Caml_obj.caml_obj_truncate(hbucket, prev_len);
      }
      if (len > t.limit && prev_len <= t.limit) {
        t.oversize = t.oversize - 1 | 0;
      }
      
    }
    t.rover = (t.rover + 1 | 0) % t.table.length;
    
  };
  var add_aux = function (t, setter, d, h, index) {
    var bucket = Caml_array.caml_array_get(t.table, index);
    var hashes = Caml_array.caml_array_get(t.hashes, index);
    var sz = bucket.length;
    var _i = 0;
    while(true) {
      var i = _i;
      if (i >= sz) {
        var newsz = Caml_primitive.caml_int_min((Caml_int32.imul(3, sz) / 2 | 0) + 3 | 0, Sys.max_array_length - 0 | 0);
        if (newsz <= sz) {
          throw {
                ExceptionID: "Failure",
                _1: "Weak.Make: hash bucket cannot grow more"
              };
        }
        var newbucket = Caml_weak.caml_weak_create(newsz);
        var newhashes = Caml_array.caml_make_vect(newsz, 0);
        Caml_weak.caml_weak_blit(bucket, 0, newbucket, 0, sz);
        $$Array.blit(hashes, 0, newhashes, 0, sz);
        Curry._3(setter, newbucket, sz, d);
        Caml_array.caml_array_set(newhashes, sz, h);
        Caml_array.caml_array_set(t.table, index, newbucket);
        Caml_array.caml_array_set(t.hashes, index, newhashes);
        if (sz <= t.limit && newsz > t.limit) {
          t.oversize = t.oversize + 1 | 0;
          for(var _i$1 = 0; _i$1 <= 2; ++_i$1){
            test_shrink_bucket(t);
          }
        }
        if (t.oversize > (t.table.length >> 1)) {
          var oldlen = t.table.length;
          var newlen = next_sz(oldlen);
          if (newlen > oldlen) {
            var newt = create(newlen);
            var add_weak = (function(newt){
            return function add_weak(ob, oh, oi) {
              var setter = function (nb, ni, param) {
                return Caml_weak.caml_weak_blit(ob, oi, nb, ni, 1);
              };
              var h = Caml_array.caml_array_get(oh, oi);
              return add_aux(newt, setter, undefined, h, get_index(newt, h));
            }
            }(newt));
            iter_weak(add_weak, t);
            t.table = newt.table;
            t.hashes = newt.hashes;
            t.limit = newt.limit;
            t.oversize = newt.oversize;
            t.rover = t.rover % newt.table.length;
            return ;
          }
          t.limit = Pervasives.max_int;
          t.oversize = 0;
          return ;
        } else {
          return ;
        }
      }
      if (Caml_weak.caml_weak_check(bucket, i)) {
        _i = i + 1 | 0;
        continue ;
      }
      Curry._3(setter, bucket, i, d);
      return Caml_array.caml_array_set(hashes, i, h);
    };
  };
  var add = function (t, d) {
    var h = Curry._1(H.hash, d);
    return add_aux(t, Caml_weak.caml_weak_set, Caml_option.some(d), h, get_index(t, h));
  };
  var find_or = function (t, d, ifnotfound) {
    var h = Curry._1(H.hash, d);
    var index = get_index(t, h);
    var bucket = Caml_array.caml_array_get(t.table, index);
    var hashes = Caml_array.caml_array_get(t.hashes, index);
    var sz = bucket.length;
    var _i = 0;
    while(true) {
      var i = _i;
      if (i >= sz) {
        return Curry._2(ifnotfound, h, index);
      }
      if (h === Caml_array.caml_array_get(hashes, i)) {
        var v = Caml_weak.caml_weak_get_copy(bucket, i);
        if (v !== undefined) {
          if (Curry._2(H.equal, Caml_option.valFromOption(v), d)) {
            var v$1 = Caml_weak.caml_weak_get(bucket, i);
            if (v$1 !== undefined) {
              return Caml_option.valFromOption(v$1);
            }
            _i = i + 1 | 0;
            continue ;
          }
          _i = i + 1 | 0;
          continue ;
        }
        _i = i + 1 | 0;
        continue ;
      }
      _i = i + 1 | 0;
      continue ;
    };
  };
  var merge = function (t, d) {
    return find_or(t, d, (function (h, index) {
                  add_aux(t, Caml_weak.caml_weak_set, Caml_option.some(d), h, index);
                  return d;
                }));
  };
  var find = function (t, d) {
    return find_or(t, d, (function (_h, _index) {
                  throw {
                        ExceptionID: "Not_found"
                      };
                }));
  };
  var find_opt = function (t, d) {
    var h = Curry._1(H.hash, d);
    var index = get_index(t, h);
    var bucket = Caml_array.caml_array_get(t.table, index);
    var hashes = Caml_array.caml_array_get(t.hashes, index);
    var sz = bucket.length;
    var _i = 0;
    while(true) {
      var i = _i;
      if (i >= sz) {
        return ;
      }
      if (h === Caml_array.caml_array_get(hashes, i)) {
        var v = Caml_weak.caml_weak_get_copy(bucket, i);
        if (v !== undefined) {
          if (Curry._2(H.equal, Caml_option.valFromOption(v), d)) {
            var v$1 = Caml_weak.caml_weak_get(bucket, i);
            if (v$1 !== undefined) {
              return v$1;
            }
            _i = i + 1 | 0;
            continue ;
          }
          _i = i + 1 | 0;
          continue ;
        }
        _i = i + 1 | 0;
        continue ;
      }
      _i = i + 1 | 0;
      continue ;
    };
  };
  var find_shadow = function (t, d, iffound, ifnotfound) {
    var h = Curry._1(H.hash, d);
    var index = get_index(t, h);
    var bucket = Caml_array.caml_array_get(t.table, index);
    var hashes = Caml_array.caml_array_get(t.hashes, index);
    var sz = bucket.length;
    var _i = 0;
    while(true) {
      var i = _i;
      if (i >= sz) {
        return ifnotfound;
      }
      if (h === Caml_array.caml_array_get(hashes, i)) {
        var v = Caml_weak.caml_weak_get_copy(bucket, i);
        if (v !== undefined) {
          if (Curry._2(H.equal, Caml_option.valFromOption(v), d)) {
            return Curry._2(iffound, bucket, i);
          }
          _i = i + 1 | 0;
          continue ;
        }
        _i = i + 1 | 0;
        continue ;
      }
      _i = i + 1 | 0;
      continue ;
    };
  };
  var remove = function (t, d) {
    return find_shadow(t, d, (function (w, i) {
                  return Caml_weak.caml_weak_set(w, i, undefined);
                }), undefined);
  };
  var mem = function (t, d) {
    return find_shadow(t, d, (function (_w, _i) {
                  return true;
                }), false);
  };
  var find_all = function (t, d) {
    var h = Curry._1(H.hash, d);
    var index = get_index(t, h);
    var bucket = Caml_array.caml_array_get(t.table, index);
    var hashes = Caml_array.caml_array_get(t.hashes, index);
    var sz = bucket.length;
    var _i = 0;
    var _accu = /* [] */0;
    while(true) {
      var accu = _accu;
      var i = _i;
      if (i >= sz) {
        return accu;
      }
      if (h === Caml_array.caml_array_get(hashes, i)) {
        var v = Caml_weak.caml_weak_get_copy(bucket, i);
        if (v !== undefined) {
          if (Curry._2(H.equal, Caml_option.valFromOption(v), d)) {
            var v$1 = Caml_weak.caml_weak_get(bucket, i);
            if (v$1 !== undefined) {
              _accu = /* :: */[
                Caml_option.valFromOption(v$1),
                accu
              ];
              _i = i + 1 | 0;
              continue ;
            }
            _i = i + 1 | 0;
            continue ;
          }
          _i = i + 1 | 0;
          continue ;
        }
        _i = i + 1 | 0;
        continue ;
      }
      _i = i + 1 | 0;
      continue ;
    };
  };
  var stats = function (t) {
    var len = t.table.length;
    var lens = $$Array.map((function (prim) {
            return prim.length;
          }), t.table);
    $$Array.sort(Caml_primitive.caml_int_compare, lens);
    var totlen = $$Array.fold_left((function (prim, prim$1) {
            return prim + prim$1 | 0;
          }), 0, lens);
    return /* tuple */[
            len,
            count(t),
            totlen,
            Caml_array.caml_array_get(lens, 0),
            Caml_array.caml_array_get(lens, len / 2 | 0),
            Caml_array.caml_array_get(lens, len - 1 | 0)
          ];
  };
  return {
          create: create,
          clear: clear,
          merge: merge,
          add: add,
          remove: remove,
          find: find,
          find_opt: find_opt,
          find_all: find_all,
          mem: mem,
          iter: iter,
          fold: fold,
          count: count,
          stats: stats
        };
}

var create = Caml_weak.caml_weak_create;

function length(prim) {
  return prim.length;
}

var set = Caml_weak.caml_weak_set;

var get = Caml_weak.caml_weak_get;

var get_copy = Caml_weak.caml_weak_get_copy;

var check = Caml_weak.caml_weak_check;

var blit = Caml_weak.caml_weak_blit;

export {
  create ,
  length ,
  set ,
  get ,
  get_copy ,
  check ,
  fill ,
  blit ,
  Make ,
  
}
/* No side effect */
